### PR TITLE
Avoid workspace structure for single packages

### PR DESCRIPTION
**Pull request by thefloweringash**

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -339,16 +350,27 @@ in rec {
...
-        chmod -R +w node_modules
+
+          mkdir node_modules
+        ''}
 
         ${linkDirFunction}
 
         linkDirToDirLinks "$(dirname node_modules/${pname})"
+        ${lib.optionalString (! useWorkspace) ''
+          ln -sT "../deps/${pname}" "node_modules/${pname}"
```
Comment (thefloweringash) [inline default.nix:372]: This was the line you eliminated in cc78373f359d17af2bbdf625338932c8ec851ae5

File: default.nix [hunk 2]
```diff
@@ -143,9 +145,14 @@ in rec {
       buildPhase = ''
         runHook preBuild
 
-        mkdir -p "deps/${pname}"
-        cp ${packageJSON} "deps/${pname}/package.json"
-        cp ${workspaceJSON} ./package.json
+        ${if useWorkspace then ''
```
Comment (thefloweringash) [inline default.nix:148]: When building modules, just yarn install without any `deps` structure.

File: default.nix [hunk 3]
```diff
@@ -339,16 +350,27 @@ in rec {
         mv $NIX_BUILD_TOP/temp "$PWD/deps/${pname}"
         cd $PWD
 
-        if [ -d ${deps}/deps/${pname}/node_modules ]; then
-          cp -r ${deps}/deps/${pname}/node_modules "deps/${pname}/node_modules"
+        ${if useWorkspace then ''
```
Comment (thefloweringash) [inline default.nix:353]: Preserve the existing structure entirely by moving the top-level `node_modules` into `deps/${pname}/`. if not using workspace structure, the top-level `node_modules` will be empty, except for a link to the main package in `deps/`.

File: default.nix [hunk 4]
```diff
@@ -425,14 +447,12 @@ in rec {
     buildPhase = ''
       source ${./nix/expectShFunctions.sh}
 
-      expectFilePresent ./node_modules/.yarn-integrity
-
       # check dependencies are installed
-      expectFilePresent ./node_modules/@yarnpkg/lockfile/package.json
+      expectFilePresent ./deps/yarn2nix/node_modules/@yarnpkg/lockfile/package.json
```
Comment (thefloweringash) [inline default.nix:451]: These expectations were assuming yarn2nix's dependencies being hoisted outside of the yarn2nix dir, which has been explicitly removed here.

**Review by thefloweringash — COMMENTED** [review]
This seems to have the desired effect. Checked with:
```shell
diff -ur 
  node_modules \
  $(nix-shell -A some_package --run 'echo $node_modules') \
  | filterdiff -i '*/package.json'
```

When empty, then the structure is the same.

File: default.nix (hunk 1 above)
Comment (thatsmydoing) [reply default.nix:372]: So they weren't using workspaces at some point?

**Review by thatsmydoing — APPROVED** [review]
Seems good.

If we're going to be updating yarn2nix, something I actually want to try at some point is to have a build phase where dev dependencies are available and only have production dependencies in the final package. Additionally, we  could probably also use the `-slim` variants of node for the final package as well.

File: default.nix (hunk 1 above)
Comment (thefloweringash) [reply default.nix:372]: Or workspaces didn't make symlinks at top level to constituent packages?  I'm not really sure.

**Pull request merged**